### PR TITLE
Generate drift tables

### PR DESCRIPTION
**Pull request by ikbendewilliam**

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -1,4 +1,7 @@
 # Changelog
+## [7.0.0] - 2022-12-25
+- Added support for generating drift tables by adding `generate_table: true` on model definition
```
Comment (NicolaVerbeeck) [inline CHANGELOG.md:3]: Use different name to tie this to drift

File: README.md [hunk 2]
```diff
@@ -581,4 +620,130 @@ generate these for you by setting `retrofit_compute: true` in the pubspec.yaml f
 ```yaml
 model_generator:
   retrofit_compute: true
+```
+
+## Drift table generation support
+
+By adding `generate_table: true`, a drift table will be generated aswel as the regular model. It will contain all fields and have extensions made to convert between DB and Generated model. In case a field shouldn't be added to the database, you can specify `ignore_for_table: true` on the field. This will not add the field to the table, but will still be added to the getModel method. If the field is required, it will also be required in the getModel method. Otherwise it will be optional.
```
Comment (NicolaVerbeeck) [inline README.md:627]: as well

File: bin/model_generator.dart [hunk 3]
```diff
@@ -114,6 +117,26 @@ void writeToFiles(
...
         modelGeneratorConfig,
       ).write();
+      if (model.generateTable == true) {
+        for (final field
+            in model.fields.where((element) => !element.ignoreForTable)) {
+          final fieldModel = modelGeneratorConfig.models.firstWhereOrNull(
+              (element) =>
+                  element.name == field.type.name ||
+                  (field.type is MapType &&
+                      element.name == (field.type as MapType).valueName));
```
Comment (NicolaVerbeeck) [inline bin/model_generator.dart:127]: What about lists? Sets? Keys as enum?

File: lib/model/field.dart [hunk 4]
```diff
@@ -17,6 +17,10 @@ class Field {
   final bool ignoreEquality;
   final String? defaultValue;
   final bool disallowNull;
+  final bool isTablePrimaryKey;
+  final bool tableAutoIncrement;
+  bool ignoreForTable;
```
Comment (NicolaVerbeeck) [inline lib/model/field.dart:22]: mutable model is bad for code cleanliness

File: lib/writer/drift_model_writer.dart [hunk 5]
```diff
@@ -0,0 +1,255 @@
...
+      }
+    }
+
+    for (final field in jsonModel.fields) {
+      final type = field.type;
+      if (!TypeChecker.isKnownDartType(type.name) &&
+          type.name != jsonModel.name) {
+        imports.addAll(_getImportsFromPath(type.name));
+      }
+      if (type is MapType && !TypeChecker.isKnownDartType(type.valueName)) {
```
Comment (NicolaVerbeeck) [inline lib/writer/drift_model_writer.dart:57]: Lists, sets, keys in maps

File: lib/writer/drift_model_writer.dart [hunk 6]
```diff
@@ -0,0 +1,255 @@
...
+        imports.addAll(_getImportsFromPath(type.valueName));
+      }
+    }
+    for (final field in extendsFields) {
```
Comment (NicolaVerbeeck) [inline lib/writer/drift_model_writer.dart:61]: This feels like duplicate code from the json model gen, we should make a helper for this

File: lib/writer/drift_model_writer.dart [hunk 7]
```diff
@@ -0,0 +1,255 @@
...
+    if (modelDescription != null && modelDescription.isNotEmpty) {
+      sb.writeln("///$modelDescription");
+    }
+    sb.writeln('class Db${modelNameUpperCamelCase}Table extends Table {');
+
+    for (final field in jsonModel.fields.where((element) =>
+        !element.ignoreForTable &&
+        (element.type is ArrayType || element.type is MapType))) {
+      print(
+          '[WARNING] ${field.name} is an array or map. This is not supported in tables atm and will be ignored.');
```
Comment (NicolaVerbeeck) [inline lib/writer/drift_model_writer.dart:85]: This should be an error

**Review by NicolaVerbeeck — CHANGES_REQUESTED** [review]
It should be specified that custom converters are not supported for drift (unless we add support for them)

File: CHANGELOG.md (hunk 1 above)
Comment (ikbendewilliam) [reply CHANGELOG.md:3]: I specifically didn't want this in case we (or someone else) adds a converter for another database package 🤔

Comment (NicolaVerbeeck) [reply CHANGELOG.md:3]: And then they have a name for this one. We can keep the 'primary_key' names etc to be reused.
Else we would beed a: generate tables: true, flavor: drift. Which would be the same as: generate_drift_tables

File: bin/model_generator.dart (hunk 3 above)
Comment (ikbendewilliam) [reply bin/model_generator.dart:127]: Lists, sets have the type inside type.name, but actually since we don't support maps atm in drift tables maps should not be added here

File: lib/util/model_helper.dart [hunk 8]
```diff
@@ -0,0 +1,112 @@
...
+    if (keyType is ArrayType) {
+      return 'List<${keyType.name}>$nullableFlag';
+    } else if (keyType is MapType) {
+      return 'Map<${keyType.name}, ${keyType.valueName}>$nullableFlag';
+    } else {
+      return '${keyType.name}$nullableFlag';
+    }
+  }
+
+  static Iterable<String> getImportsFromField(
```
Comment (NicolaVerbeeck) [inline lib/util/model_helper.dart:73]: Nit: can be private

**Review by NicolaVerbeeck — APPROVED** [review]

File: test/writer/drift_model_writer/primary_key_single/config.txt [hunk 9]
```diff
@@ -0,0 +1,9 @@
+Person:
+  path: user/person/
+  generate_drift_table: true
+  properties:
+    firstName:
+      type: string
+      required: true
+      is_table_primary_key: true
+    lastName: string
```
Comment (vanlooverenkoen) [inline test/writer/drift_model_writer/primary_key_single/config.txt:9]: output.txt is generating a nullable table? this is not correct.

File: test/writer/writer_test_helper.dart [hunk 10]
```diff
@@ -39,4 +40,25 @@ class WriterTestHelper {
...
+        File('test/writer/drift_model_writer/$resultFileName/config.txt');
+    final output = file.readAsStringSync();
+    final pubspecContent = pubspecFile.readAsStringSync();
+    final configContent = configFile.readAsStringSync();
+    final pubspecConfig = PubspecConfig(pubspecContent);
+    final ymlConfig = YmlGeneratorConfig(pubspecConfig, configContent, '');
+    final actual = DriftModelWriter(
+            pubspecConfig, model, extendsFields, enumFields, ymlConfig)
+        .write();
+    // print(actual);
```
Comment (vanlooverenkoen) [inline test/writer/writer_test_helper.dart:61]: remove comment?

**Review by vanlooverenkoen — CHANGES_REQUESTED** [review]
Not sure if all the test cover all cases. We should maybe create a list of edgecases and check if the tests are covering those edge cases as well

File: README.md [hunk 11]
```diff
@@ -581,4 +620,132 @@ generate these for you by setting `retrofit_compute: true` in the pubspec.yaml f
...
+    pages: int?
+    isAvailable: bool
+    authors:
+      type: array
+      required: true # this is required, so will be added to the getModel method as required
+      ignore_for_table: true # You can specify whether a field is added to the table or not. If it is not added, you can still specify it in the getModel method to convert between DB model and generated model
+      items:
+        type: Person
+    publishers:
+      type: array # Arrays will not be added to the table, but specified in the getModel method. You should add another table for them (manually or generated)
```
[the remark this answers is not in the document]
Comment (ikbendewilliam) [reply README.md:656]: For now you can have a table generated from the other model and add a field somethingId and have a dao that joins them together.
I can also have a look to generate a converter to support lists of dart types (num/string/bool/datetime) or support using toJson.
WDYT?

File: test/writer/drift_model_writer/primary_key_single/config.txt (hunk 9 above)
Comment (ikbendewilliam) [reply test/writer/drift_model_writer/primary_key_single/config.txt:9]: We need to fix #126, I'll do that first 🙈

File: test/writer/drift_model_writer/specified_database_path/output.txt [hunk 12]
```diff
@@ -0,0 +1,24 @@
+// GENERATED CODE - DO NOT MODIFY BY HAND
+
+import 'package:drift/drift.dart';
+import 'package:model_generator_example/base_dir/path_to_my_model/person.dart';
+import 'package:model_generator_example/database/example.dart';
+
+@DataClassName('DbPerson')
+class DbPersonTable extends Table {
+  ///A good description
+  TextColumn get firstName => text().nullable()();
```
[the remark this answers is not in the document]
Comment (ikbendewilliam) [reply test/writer/drift_model_writer/specified_database_path/output.txt:10]: No, it's not required so it can be null :-) 
```yaml
firstName:
  description: A good description
  type: string
```

File: test/writer/drift_model_writer/enum_field/output.txt [hunk 13]
```diff
@@ -0,0 +1,45 @@
...
+      );
+}
+
+class PersonTableGenderConverter extends TypeConverter<Gender, String> {
+  const PersonTableGenderConverter();
+
+  @override
+  Gender fromSql(String fromDb) {
+    for (final value in Gender.values) {
+      if (value.toString() == fromDb) return value;
```
[the remark this answers is not in the document]
Comment (ikbendewilliam) [reply test/writer/drift_model_writer/enum_field/output.txt:36]: Good point, this can be solved if we did something else I'd like to see. I suggest we add the JSONValue to the enum as a field, this way we can much more easily convert between the two and can also use that value to do the check here :-) WDYT

[the remark this answers is not in the document]
Comment (ikbendewilliam) [reply test/writer/drift_model_writer/enum_field/output.txt:36]: Fixed in #129

File: test/writer/drift_model_writer/primary_key_single/config.txt (hunk 9 above)
Comment (ikbendewilliam) [reply test/writer/drift_model_writer/primary_key_single/config.txt:9]: Fixed in #128

**Comment by vanlooverenkoen** [comment]
`generate_drift_table` `ignore_for_table` I would use `ignore_for_drift_table` here as well.

**Comment by vanlooverenkoen** [comment]
```
 authors is an array or map. Ignore this field by adding ignore_for_table: true
```

improve error handling.

```
`authors` is an "array or maps", this is not support to save in a table. You can ignore this field by adding

   ignore_for_table: true

```

Also, specify if it is a map or array. We already know this.

**Comment by vanlooverenkoen** [comment]
There is no error thrown when I create a table. without a primary key.
Maybe we don't need an error, but we should probably log this.

**Comment by vanlooverenkoen** [comment]
Question: what happens if we want to add an extra field that is not returned from the API? And we don't want to be creating an extra table, and add complexity of doing a lookup in that extra table?

**Comment by vanlooverenkoen** [comment]
When using a `type: custom` we should allow to specify a `TypeConverter`. Right now it just said, `No drift column or type for Duration` but duration could be mapped to the amount of seconds or miliseconds.

**Comment by vanlooverenkoen** [comment]
Tables generate 1 enter to much.  after all columns

**Comment by vanlooverenkoen** [comment]
There is 1 big bug I thinks with enums.

This is the config
```
    secondCategory: BookCategory?
```

This is generated. 
```
class BookTableBookCategoryConverter extends TypeConverter<BookCategory, String> {
  const BookTableBookCategoryConverter();

  @override
  BookCategory fromSql(String fromDb) {
    for (final value in BookCategory.values) {
      if (value.jsonValue == fromDb) return value;
    }
    return BookCategory.values.first;
  }

  @override
  String toSql(BookCategory value) {
    return value.jsonValue;
  }
}
```

I would expect the BookCategory in the `BookTableBookCategoryConverter` to be nullable. Becuase right now you just return the first item. Which will result in incorrectly saved values.

**Comment by ikbendewilliam** [comment]
> A generated table imports
> 
> ```
> import 'package:flutter/foundation.dart';
> ```
> 
> but it is never needed.
> 
> config
> 
> ```
> Book2:
>   path: book/
>   generate_drift_table: true
>   properties:
>     id:
>       type: int
>       required: true
>       is_table_primary_key: true
>     name: string
> ```

Note that this is added in the config in pubspec.yaml
```
  extra_imports:
    - 'package:flutter/foundation.dart'
```

**Comment by ikbendewilliam** [comment]
> Question: what happens if we want to add an extra field that is not returned from the API? And we don't want to be creating an extra table, and add complexity of doing a lookup in that extra table?
I'll add `only_for_table` for this

**Comment by ikbendewilliam** [comment]
> When using a type: custom we should allow to specify a TypeConverter. Right now it just said, No drift column or type for Duration but duration could be mapped to the amount of seconds or miliseconds.

I've added extra_imports_for_table and type_converter_for_table for this

**Comment by ikbendewilliam** [comment]
Added a nullable enum converter for non-required fields

**Comment by vanlooverenkoen** [comment]
Decline for now. Can be revisited in the future
